modules/ui/rapid_feature_toggle_dialog.js

```diff
@@ -1,192 +1,323 @@
 
-import { t } from '../core/localizer';
+import marked from 'marked';
+import { t, localizer } from '../core/localizer';
+import { prefs } from '../core/preferences';
 import { icon } from './intro/helper';
+import { svgIcon } from '../svg/icon';
 import { uiModal } from './modal';
-import marked from 'marked';
-
+import { uiRapidColorpicker } from './rapid_colorpicker';
+import { uiRapidViewManageDatasets } from './rapid_view_manage_datasets';
 
 export function uiRapidFeatureToggleDialog(context, AIFeatureToggleKey, featureToggleKeyDispatcher) {
+  const rapidContext = context.rapidContext();
+  const showPreview = prefs('rapid-internal-feature.esriPreview') === 'true';
 
+  let _modalSelection = d3_select(null);
+  let _content = d3_select(null);
+  let _viewManageModal;
+  let _colorpicker;
 
-    function toggleSvgBuildings () {
-        var drawAiFeatures = context.layers().layer('ai-features');
-        drawAiFeatures.toggleBuildings();
-    }
 
+  function datasetEnabled(d) {
+    const dataset = rapidContext.datasets()[d.id];
+    return dataset && dataset.enabled;
+  }
 
-    function toggleSvgRoads() {
-        var drawAiFeatures = context.layers().layer('ai-features');
-        drawAiFeatures.toggleRoads();
+  function toggleDataset(d) {
+    const dataset = rapidContext.datasets()[d.id];
+    if (dataset) {
+      dataset.enabled = !dataset.enabled;
+      context.map().pan([0,0]);   // trigger a map redraw
+    }
+  }
+
+  function changeColor(datasetID, color) {
+    const dataset = rapidContext.datasets()[datasetID];
+    if (dataset) {
+      dataset.color = color;
+      context.map().pan([0,0]);   // trigger a map redraw
+      _content.call(renderModalContent);
+
+      // if a RapiD feature is selected, reselect it to update sidebar too
+      const mode = context.mode();
+      if (mode && mode.id === 'select-ai-features')
+      context.enter(mode, mode.selectedDatum());
     }
+  }
 
+  function toggleRapid() {
+    const rapidLayer = context.layers().layer('ai-features');
+    rapidLayer.enabled(!rapidLayer.enabled());   // toggling the layer will trigger a map redraw
+    _content.call(renderModalContent);
+  }
 
-    function handleToggleAllClick(){
-        var drawAiFeatures = context.layers().layer('ai-features');
-        drawAiFeatures.enabled(!drawAiFeatures.enabled());
-        redrawOnToggle();
+
+  function keyPressHandler() {
+    if (d3_event.shiftKey && d3_event.key === t('map_data.layers.ai-features.key')) {
+      toggleRapid();
     }
+  }
 
 
-    function keyPressFormHandler(){
-        if (d3_event.shiftKey &&
-            d3_event.key === t('map_data.layers.ai-features.key')){
-            handleToggleAllClick();
-        }
-    }
+  return function render(selection) {
+    _modalSelection = uiModal(selection);
 
+    _modalSelection.select('.modal')
+      .attr('class', 'modal rapid-modal');   // RapiD styling
 
-    function redrawOnToggle() {
```

Comment:
I particularly like the addition of the enter/exit() clauses for each separate layer, just so we don't have to come up with `n` different css 'hide' classes for each layer the user wants to see!